src/libs/swaps/basis/rBasisSwap.ts

```diff
@@ -0,0 +1,613 @@
+    accountInfo.delegate = null;
+    accountInfo.delegatedAmount = new u64(0);
+  } else {
+    accountInfo.delegate = new PublicKey(accountInfo.delegate);
+    accountInfo.delegatedAmount = u64.fromBuffer(accountInfo.delegatedAmount);
+  }
+
+  accountInfo.isInitialized = accountInfo.state !== 0;
+  accountInfo.isFrozen = accountInfo.state === 2;
+
+  if (accountInfo.isNativeOption === 1) {
+    accountInfo.rentExemptReserve = u64.fromBuffer(accountInfo.isNative);
+    accountInfo.isNative = true;
+  } else {
+    accountInfo.rentExemptReserve = null;
+    accountInfo.isNative = false;
+  }
+
+  if (accountInfo.closeAuthorityOption === 0) {
+    accountInfo.closeAuthority = null;
+  } else {
+    accountInfo.closeAuthority = new PublicKey(accountInfo.closeAuthority);
+  }
+
+  return accountInfo;
+}
+
+export async function getTokenAccount(connection, publicKey) {
+  const result = await connection.getAccountInfo(publicKey);
+  if (!result) return false;
+  const data = Buffer.from(result.data);
+  const account = parseTokenAccountData(publicKey, data);
+  return {
+    publicKey,
+    account,
+  };
+}
+
+export const unstakeBasisInstruction = (
+  amount: number | BN | string,
+  userAuthority: PublicKey,
+  userToken: PublicKey,
+  userRedeemable: PublicKey,
+): TransactionInstruction => {
+  const dataLayout = BufferLayout.struct([
+    Layout.uint64('amount'),
+  ]);
+
+  const data = Buffer.alloc(dataLayout.span);
+  dataLayout.encode(
+    {
+      amount: new BN(amount),
+    },
+    data,
+  );
+
+  // userAuthority    M[ ] S[x]
+  // userToken        M[x] S[ ]
+  // userRedeemable   M[x] S[ ]
+  // redeemableMint   M[x] S[ ]
+  // tokenMint        M[ ] S[ ]
+  // stakingAccount   M[ ] S[ ]
+  // tokenVault       M[x] S[ ]
+  // tokenProgram     M[ ] S[ ]
+
+  const keys = [
+    { pubkey: userAuthority, isSigner: true, isWritable: false },
+    { pubkey: userToken, isSigner: false, isWritable: true },
+    { pubkey: userRedeemable, isSigner: false, isWritable: true },
+    { pubkey: new PublicKey(MINT_RBASIS), isSigner: false, isWritable: true },
+    { pubkey: new PublicKey(MINT_BASIS), isSigner: false, isWritable: false },
+    { pubkey: new PublicKey(PROGRAM_BASIS_STAKING_INSTANCE), isSigner: false, isWritable: false },
+    { pubkey: new PublicKey(PROGRAM_BASIS_STAKING_VAULT), isSigner: false, isWritable: true },
+    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
+  ];
+
+  return new TransactionInstruction({
+    keys,
+    programId: new PublicKey(PROGRAM_BASIS_STAKING),
+    data: Buffer.concat([sighash(SIGHASH_GLOBAL_NAMESPACE, StakingInstructionNames[StakingInstruction.unstake]), data]),
+  });
+};
+
+export const unstakeBasis = async (
+  connection: Connection,
+  payer: Account,
+) => {
+  const ixs: TransactionInstruction[] = [];
+
+  // get associated token account for Basis (or create if doens't exist)
+  const BasisAccount = await Token.getAssociatedTokenAddress(
```

Comment:
can save compute by moving this to after we check for rbasis amount in wallet and only fetch/create ata when needed